ug: Add doc on a custom CI for Apps

mike-sul opened the pull request:
Signed-off-by: Mike <mike.sul@foundries.io>

## Readiness

* [x] Merge (pending reviews)
* [ ] Merge after _date or event_
* [ ] Draft

## Overview

_Why merge this PR? What does it solve?_

## Checklist

_Optional. Add a 'x' to steps taken._
_You can fill this out after opening the PR. "Did I..."_

* [X] Run spelling and grammar check, preferably with linter.
* [ ] Avoid changing any header associated with a link/reference.
* [x] Step through instructions (or ask someone to do so).
* [ ] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [x] Match tone and style of page/section.
* [X] Run `make linkcheck`.
* [x] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [x] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [ ] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [x] Request PR review by a technical writer and at least one peer.

## Comments

_Any thing else that a maintainer/reviewer should know._
_This could include potential issues, rational for approach, etc._

doanac commented:
Docs for 723c2aae60db2c245c9322db4f4bcfc4bb6e10e2 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1827/docs/artifacts/html/index.html

mike-sul commented:
@doanac I wonder if I should mention the following two approaches in this guide too (keeping source code in your own repo and still use Fio CI).

1. https://github.com/foundriesio/gh-action-update-submodule
2. https://github.com/foundriesio/mirror-action

doanac commented:
Docs for a835979603f464a67bce9dc224b3e766a92ce2b8 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1828/docs/artifacts/html/index.html

doanac commented:
nice start. i'll defer to katrina on those links. I'm not sure they fit well in the scope of this, but maybe we can put some "notes" at the end of this.

mike-sul commented:
> but maybe we can put some "notes" at the end of this.

Yep, that's exactly what I thought, just `Note` at the end. Adding them within the guide context will just confuse a reader.

doanac commented:
Docs for c034dfe1c78295d97b761f73d2ac2632b6e38365 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1829/docs/artifacts/html/index.html

angolini commented on source/user-guide/custom-ci/custom-ci.rst:
```diff
@@ -0,0 +1,149 @@
+Prerequisites
+-------------
+
+1. Your Factory has been successfully created.
+
+2. At least one successful Factory CI build and corresponding Target with the tag and a hardware ID that you will be using in the following guide.
+
+3. You have a GitHub repo with source code, Dockerfiles and a Docker compose file that work.
+
+For example, the prerequisites for the following guide are:
+
+1. Factory ``lmp-demo`` is setup and has been successfully created.
+2. The successfully built Target with the tag ``custom-ci-devel`` and the hardware ID ``raspberrypi4-64``.
```
```suggestion
1. Your Factory has been successfully created.

2. At least one successful Factory CI build and the corresponding Target with the tag and a hardware ID you will use in the following guide.

3. You have a GitHub repo with source code, Dockerfiles, and a Docker compose file that work.

For example, the prerequisites for the following guide are:

1. Factory ``lmp-demo`` is setup and has been successfully created.
2. The successfully built Target with the tag ``custom-ci-devel`` and the hardware ID ``raspberrypi4-64``.
```
suggestion automatically generated by the app I use to review the text

mike-sul replied:
fixed

angolini commented on source/user-guide/custom-ci/custom-ci.rst:
```diff
@@ -0,0 +1,149 @@
+1. Pins images referred from the App compose file.
+    a) If an image is already pinned (digest, a reference with sha256 hash) it does nothing.
+    b) If an image is referred by a tag it tries to get the image digest - a reference with sha256 hash. If it fails to obtain an image digest then the utility exists with an error.
+    c) If an image reference has not tag nor hash it checks if it's specified via ``--pinned-images`` input parameter. If no digest reference found in ``pinned-images`` the the utility exists with an error.
```
```suggestion
    c) If an image reference has not tag nor hash it checks if it's specified via ``--pinned-images`` input parameter. If no digest reference is found in ``pinned-images`` the utility exists with an error.
```

mike-sul replied:
fixed

angolini commented on source/user-guide/custom-ci/custom-ci.rst:
```diff
@@ -0,0 +1,149 @@
+The utility outputs the built and pushed App image digest to the file specified via ``-d`` input parameter.
+Then the published App can be referenced with a hashed URI - ``hub.foundries.io/<factory>/<app-name>@sha256:<hash>``.
+
+Once, the App in successfully built and pushed to the registry, a new Target referring to it can be created.
```
```suggestion
Once the App is successfully built and pushed to the registry, a new Target referring to it can be created.
```

mike-sul replied:
fixed

doanac commented:
Docs for 1287d6ef285a9c7b08dffcab1c6196709d87d53c are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1833/docs/artifacts/html/index.html

mike-sul commented on source/user-guide/custom-ci/custom-ci.rst:
```diff
@@ -0,0 +1,149 @@
+______________________________
+The next step is to define a GitHub actions workflow in your repo or extend an existing one.
+`The sample GitHub actions workflow`_ provides an example of the workflow that communicates with FoundiresFactory to achieve the goal,
+i.e. the items listed at the end of the first paragraph.
```
No I meant 

```
This section guides you through the steps of creating a custom CI pipeline in GitHub that:

- builds multi-arch container images and pushes them to `Foundries Registry`_;
- builds `Foundries Compose App`_ and pushes it `Foundries Registry`_;
- composes `TUF Targets role metadata`_ compliant with FoundriesFactory TUF requirements;
- adds the composed TUF Targets to `FoundriesFactory Targets`_.
```

I suppose this is "the end of the introduction section"?

mike-sul commented on source/user-guide/custom-ci/custom-ci.rst:
```diff
@@ -0,0 +1,149 @@
+- builds multi-arch container images and pushes them to `Foundries Registry`_;
+- builds `Foundries Compose App`_ and pushes it `Foundries Registry`_;
+- composes `TUF Targets role metadata`_ compliant with FoundriesFactory TUF requirements;
+- adds the composed TUF Targets to `FoundriesFactory Targets`_.
```
fixed. What's the app, btw? Is it grammarly?

mike-sul commented:
> Personally, I wouldn't say I like how the links to the source code look in the middle of the text. I would suggest adding them as footnotes, perhaps.

I removed the links to the source code, in some places I replaced them with an example.

doanac commented:
Docs for 371e58bd1501baf0de71a20ac324423b64d115b1 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1835/docs/artifacts/html/index.html

doanac commented:
Docs for d29e20ca27f18c6412cf2ababfd4ab374fb70ed9 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1836/docs/artifacts/html/index.html

doanac commented:
Docs for a7a76cb6db8283cadd2458569046ae757034c0f9 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1837/docs/artifacts/html/index.html

doanac commented:
Docs for fd81e2c18d9ff237c7484cb27c1963c364bad393 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1840/docs/artifacts/html/index.html

doanac commented on source/user-guide/custom-ci/custom-ci-for-apps.rst:
```diff
@@ -0,0 +1,150 @@
+- builds multi-arch container images and pushes them to `Foundries Registry`_;
+- builds `Foundries Compose App`_ and pushes it `Foundries Registry`_;
+- composes `TUF Targets role metadata`_ compliant with FoundriesFactory TUF requirements;
+- adds the composed TUF Targets to `FoundriesFactory Targets`_.
```
there's a mix of semi-colon and periods for the bulleted endings. probably one or the other or no ending. katrina will know

doanac commented on source/user-guide/custom-ci/custom-ci-for-apps.rst:
```diff
@@ -0,0 +1,150 @@
+The GitHub action needs to authenticate itself at the FoundriesFactory OTA service and the `Foundries Registry`_.
+Therefore, an access token must be created and added to the GitHub repo action tokens prior to running any CI pipelines.
+You can create the token at `FoundriesFactory WebApp`_.
+The token must have `containers:read-update` and `targets:read-update` scopes to access the registry and the OTA service correspondingly.
```
I think you may have meant to double-backtick the scopes?

kprosise replied:
@mike-sul use double. Single is for interpreted text.

kprosise commented on source/user-guide/custom-ci/custom-ci-for-apps.rst:
```diff
@@ -0,0 +1,150 @@
+You can create the token at `FoundriesFactory WebApp`_.
+The token must have `containers:read-update` and `targets:read-update` scopes to access the registry and the OTA service correspondingly.
+
+Set Token In GitHub Repo
```
```suggestion
Set Token in GitHub Repo
```

mike-sul replied:
updated

mike-sul commented on source/user-guide/custom-ci/custom-ci-for-apps.rst:
```diff
@@ -0,0 +1,150 @@
+
+2. Creates the compose App container image.
+    a) Creates an archive (``tgz``) that contains the App compose file and its supplementary files.
+    b) Creates a container image manifest referring to the App archive as an image layer/blob.
```
done

mike-sul commented:
Closed this PR in favor of https://github.com/foundriesio/docs/pull/461

mike-sul commented:
Closing this PR in favor of https://github.com/foundriesio/docs/pull/461

kprosise reviewed (commented):
Unfinished review, noting that the PR has been closed. I can finish the review if reopened.

mike-sul commented:
> Unfinished review, noting that the PR has been closed. I can finish the review if reopened.

OK, I'll reopen the PR. Once, review is completed I'll close it again and merged the final version to the new PR https://github.com/foundriesio/docs/pull/461

mike-sul commented on source/user-guide/custom-ci/custom-ci-for-apps.rst:
```diff
@@ -0,0 +1,150 @@
+
+.. _ug-custom-ci-app:
+
+Custom CI To Build Compose App Targets
```
done

doanac commented:
Docs for 9b87310de423d2831ab9c33f101eb81da0ff4f56 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1856/docs/artifacts/html/index.html

mike-sul commented:
@kprosise Updated the PR according to your comments.

mike-sul commented:
@kprosise I updated the main PR (461) according to your comments too. So, I think we can close this PR and continue reviewing in https://github.com/foundriesio/docs/pull/461?

kprosise commented:
> @kprosise I updated the main PR (461) according to your comments too. So, I think we can close this PR and continue reviewing in #461?

sure!

mike-sul commented:
Closing this PR in favor of https://github.com/foundriesio/docs/pull/461
